libraries/AC_Avoidance/AP_OABendyRuler.cpp

```diff
@@ -144,11 +192,57 @@ bool AP_OABendyRuler::update(const Location& current_loc, const Location& destin
     destination_new.offset_bearing(chosen_bearing, distance_to_dest);
 
     // log results
-    AP::logger().Write_OABendyRuler(true, chosen_bearing, best_margin, destination, destination_new);
+    AP::logger().Write_OABendyRuler(true, chosen_bearing, false, best_margin, destination, destination_new);
 
     return true;
 }
 
+/*
+This function is called when BendyRuler has found a bearing which is obstacles free at atleast lookahead_step1_dist and  then lookahead_step2_dist from the present location
+In many situations, this new bearing can be either left or right of the obstacle, and BendyRuler can have a tough time deciding between the two.
+It has the tendency to move the vehicle back and forth, if the margin obtained is even slightly better in the newer iteration.
+Therefore, this method attempts to avoid changing direction of the vehicle by more than _bendy_angle degrees, 
+unless the new margin is atleast _bendy_ratio times better than the margin with previously calculated bearing.
+We return true if we have resisted the change and will follow the last calculated bearing. 
+*/
+bool AP_OABendyRuler::avoid_bearing_change(const Location &destination, const Location &current_loc, const bool &active, const float &bearing_test, const float &lookahead_step1_dist, const float &margin, float &final_bearing, float &final_margin) 
+{      
+    bool resisted_change = false;
+    //see if there was a change in destination, if so, do not resist changing direction 
+    bool dest_change = false;
+    if(!destination.same_latlon_as(_destination_prev)) {
+        dest_change = true;
+        _destination_prev = destination;
+    }
+                        
+    //check if we need to resist the change in direction of the vehicle. If we have a clear path to destination, go there any how  
+    if (active && !dest_change && is_positive(_bendy_ratio)) { 
+        //check the change in bearing between freshly calculated and previous stored BendyRuler bearing
+        if ((fabsf(wrap_180(_stored_bearing-bearing_test)) > _bendy_angle) && (!is_equal(_stored_bearing,FLT_MAX))) {
+            // check margin in last bearing's direction
+            Location test_loc_previous_bearing = current_loc;
+            test_loc_previous_bearing.offset_bearing(wrap_180(_stored_bearing), lookahead_step1_dist);
+            float previous_bearing_margin = calc_avoidance_margin(current_loc,test_loc_previous_bearing);
+
+            if (margin < (_bendy_ratio * previous_bearing_margin)) {
+                //don't change direction abruptly. If margin difference is not significant, follow the last direction
+                final_bearing = _stored_bearing;
+                final_margin  = previous_bearing_margin;
+                resisted_change = true;
+            } 
+        } 
+    } else {
+        //reset stored bearing if BendyRuler is not active or if WP has changed for unnecessary resistance to path change
+        _stored_bearing = FLT_MAX;
```

Comment:
hmm.. this function has the side effect of updating the _stored_bearing variable.  We pass in lots of other arguments but not this one.  I think we should either make this method a const and pass in everything it needs or not pass in hardly anything and have the function pull what it needs from the global variables.

Reply:
I am now passing everything as arguments.... But the problem is I can't make this method a const because we call a non-const method `calc_avoidance_margin()` with our new method.

Reply:
@rishabsingh3003, it looks to me like all these functions can be made const.  It was probably an oversight that they weren't made const initially.

- calc_margin_from_circular_fence
- calc_margin_from_object_database
- calc_margin_from_inclusion_and_exclusion_polygons
- calc_margin_from_inclusion_and_exclusion_circles

Reply:
Done!